sklearn/base.py

```diff
@@ -377,6 +378,44 @@ def _check_n_features(self, X, reset):
                                        self.n_features_in_)
                 )
 
+    def _check_feature_names(self, X, reset=True):
+        """Validate feature names and set or check the `feature_names_in_`
+        attribute.
+
+        Parameters
+        ----------
+        X : {dataframe-like} of shape (n_samples, n_features)
+            The input samples.
+        reset : bool, default=True
+            Whether to reset the `feature_names_in_` attribute.
+            If False, the Input will be checked for consistency with data
+            provided when reset was last True.
+        """
+
+        if reset:
+            self.feature_names_in_ = _get_feature_names(X)
+            return
+
+        fitted_feature_names = getattr(self, "feature_names_in_", None)
+        if fitted_feature_names is None:
```

Comment:
do we want to raise if either `(fitted_feature_names is None xor feature_names_in is None) == True`? I think we should

Reply:
This means, we would start warning for training on numpy array and predicting on pandas dataframe. What would be a good message?

1. train on numpy and predict on dataframe: Convert your dataframe into a numpy array before predicting?
2. train on dataframe and predict on numpy array: Convert your numpy array into a dataframe before predicting?

Reply:
Maybe it would be enough to mention that the type of the array in `fit` is different than the one at `predict` and we expect a consistent type between `fit` and `predict`.

Reply:
Actually I don't know if we should raise an error or only a warning mentioning that no information about the columns would be available?

Reply:
I'm ambivalent about a warning. On the one hand it seems dangerous because it means you can't check consistency, but on the other hand people will ignore warning if there's too many. I think +0.5 for the warning?

Reply:
> Maybe it would be enough to mention that the type of the array in fit is different than the one at predict and we expect a consistent type between fit and predict.

If we were to support xarray in the future, I think we can strictly work on type. It would have to be "are there feature names in fit" and "are there feature names in predict".

Updated the PR with the warnings.